ql/src/java/org/apache/hadoop/hive/ql/udf/ptf/ValueBoundaryScanner.java

```diff
@@ -414,16 +423,35 @@ protected int computeStartPreceding(int rowIdx, PTFPartition p) throws HiveExcep
       return r + 1;
     }
     else { // Use Case 5.
+      Pair<Integer, Object> start = binaryPreSearchBack(r, p, sortKey, rowVal, amt);
+      //start again with linear search from the last point where !isDistanceGreater was true
+      r = start.getLeft();
+      rowVal = start.getRight();
       while (r >= 0 && !isDistanceGreater(sortKey, rowVal, amt) ) {
         Pair<Integer, Object> stepResult = skipOrStepBack(r, p);
         r = stepResult.getLeft();
         rowVal = stepResult.getRight();
       }
-
       return r + 1;
     }
   }
 
+  private Pair<Integer, Object> binaryPreSearchBack(int r, PTFPartition p, Object sortKey,
```

Comment:
I guess existing PTF tests should covert this optimization but would be great if we would add specific ones for cases 4 and  5 above

Reply:
makes sense, this is a WIP patch, I'll still need to add at least binaryPreSearchForward for forward searching